compiler/sem/sigmatch.nim

```diff
@@ -1597,88 +1593,70 @@ typeRel can be used to establish various relationships between types:
-        # var root = internalPut[int, int](nil, 312, 312, oldvalue)
-        # var it1 = internalFind(root, 312) # cannot instantiate: 'D'
-        #
-        # we steal the generic parameters from the tyGenericBody:
-        for i in 1..<f.len:
-          let x = PType(idTableGet(c.bindings, genericBody[i-1]))
-          if x.isNil():
-            discard "maybe fine (for e.g. a==tyNil)"
-          else:
-            c.c.graph.config.internalAssert(
-              x.kind notin {tyGenericInvocation, tyGenericParam},
-              "wrong instantiated type!")
-            let key = f[i]
-            let old = PType(idTableGet(c.bindings, key))
-            if old.isNil():
-              put(c, key, x)
-            elif typeRel(c, old, x, flags + {trDontBind}) == isNone:
-              return isNone
-
-      if result == isNone:
-        # Here object inheriting from generic/specialized generic object
-        # crossing path with metatypes/aliases, so we need to separate them
-        # by checking sym.id
-        let genericSubtype = isGenericSubtype(c, x, f, depth, f)
-        if not (genericSubtype and aobj.sym.id != fobj.sym.id) and aOrig.kind != tyGenericBody:
-          depth = -1
-
-      if depth >= 0:
+    elif body.skipTypes({tyPtr, tyRef}).kind == tyObject and
+         x.skipToObject(sp) != nil:
+      # if its not the same object type, there can still be a subtype
+      # relationship
+      let depth = isSubtypeOfGenericInstance(c, x, f, f)
+      case depth
+      of -1:
+        result = isNone
+      of 0:
+        unreachable("already handled by the ``x.kind == tyGenericInst`` branch")
+      else:
         c.inheritancePenalty += depth
-        # bug #4863: We still need to bind generic alias crap, so
-        # we cannot return immediately:
-        result = if depth == 0: isGeneric else: isSubtype
+        result = isSubtype
+
+    elif x.kind == tyGenericInst and
+         body.kind notin {tyAnd, tyOr, tyGenericInvocation}:
+      # the formal invocation is not a generic alias and both `f` and
+      # `a` are not applications to the same generic type -> no match.
+      result = isNone
+    else:
+      # XXX: to not ignore phantom types, this branch should only be taken
+      #      when `f` is not a phantom type
+      # match against the generic body, but before doing so, replace the type
+      # variables therein with the arguments provided to the invocation. For
+      # example, given the body ``Body[T] = (T, T)`` and invocation
+      # ``Body[A]``, the resulting type would be ``(A, A)``.
+      var bindings: TIdTable
+      initIdTable(bindings)
+      # prepare the parameter-to-argument bindings
+      for i in 1..<f.len:
+        idTablePut(bindings, f.base[i-1], f[i])
+
+      # XXX: since we're forwarding the arguments here, the constraints on the
+      #      body's parameters are ignored...
+      let other = replaceTypeParamsInType(c.c, bindings, body)
```

Comment:
It's a temporary regression because, it's not unreasonable to do the check as part of the replace in the future?

Reply:
Yep, my current plan goes into that direction. In short, the idea is to, should the `typeRel` against the patched body return a match, perform a `typeRel` for all type arguments against their passed-to parameter's constraint. For that to work, `typeRel` is going to need some more improvements, which is why I haven't implemented it as part of this PR.

---

A complementary improvement could be leveraging the `tyInferred` mechanism (or an improved/extended version thereof) to infer the constraints for a generic parameter on definition:
```nim
type Generic[A: object] = (A, A)

proc f[T](x: Generic[T]) = ...
```
here `T` would be inferred to be constrained to an `object` type. This would make it possible to detect conflicting constraints on *definition* of a generic type/routine (instead of during parameter type matching).